libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxSpeechApi.kt

```diff
@@ -55,6 +66,40 @@ class MapboxSpeechApi @JvmOverloads constructor(
         }
     }
 
+    /**
+     * Given [VoiceInstructions] the method will try to generate the
+     * voice instruction [SpeechAnnouncement] including the synthesized speech mp3 file
+     * from Mapbox's API Voice.
+     * NOTE: this method will NOT try downloading an mp3 file from server. It will either use
+     * an already predownloaded file or an onboard speech synthesizer. Only invoke this method
+     * if you use voice instructions predownloading (see [VoiceInstructionsPrefetcher]),
+     * otherwise invoke [generatePredownloaded] in your [VoiceInstructionsObserver].
+     * @param voiceInstruction VoiceInstructions object representing [VoiceInstructions]
+     * @param consumer is a [SpeechValue] including the announcement to be played when the
+     * announcement is ready or a [SpeechError] including the error information and a fallback
+     * with the raw announcement (without file) that can be played with a text-to-speech engine.
+     * @see [cancel]
+     */
+    fun generatePredownloaded(
+        voiceInstruction: VoiceInstructions,
+        consumer: MapboxNavigationConsumer<Expected<SpeechError, SpeechValue>>
+    ) {
+        mainJobController.scope.launch {
```

Comment:
```suggestion
```
Do you have suspend function here?

Reply:
I don't see much difference between acquiring lock from from main thread directly and from a coroutine with main dispatcher. In both cases if lock is locked, main thread will be blocked, won't it?

Reply:
You're right. It's not mutex.
I removed the lock and made sure the map is only accessed from the main thread and there are no suspensions between the lines that should be used together.